contracts/eosiolib/asset.hpp

```diff
@@ -23,22 +25,203 @@ namespace eosio {
+
+         return result;
+      }
+
+      void print()const {
+         int64_t p = (int64_t)symbol.precision();
+         int64_t p10 = 1;
+         while( p > 0  ) {
+            p10 *= 10; --p;
+         }
+         p = (int64_t)symbol.precision();
+
+         char fraction[p+1];
+         fraction[p] = '\0';
+         auto change = amount % p10;
+
+         for( int64_t i = p -1; i >= 0; --i ) {
+            fraction[i] = (change % 10) + '0';
+            change /= 10;
+         }
+         printi( amount / p10 );
+         prints(".");
+         prints_l( fraction, uint32_t(p) );
+         prints(" ");
+         symbol.print(false);
+      }
+
+      asset& operator+=( const asset& a ) {
+         eosio_assert( symbol == a.symbol, "type missmatch" );
+         return *this = (*this + a);
+      }
+
+      EOSLIB_SERIALIZE( asset, (amount)(symbol) )
+   };
+
+   struct extended_asset : public asset {
+      account_name contract;
+
+      extended_symbol get_extended_symbol()const { return extended_symbol( symbol, contract ); }
+      extended_asset(){}
+      extended_asset( int64_t v, extended_symbol s ):asset(v,s),contract(s.contract){}
+      extended_asset( asset a, account_name c ):asset(a),contract(c){}
 
-      template<typename DataStream>
-      friend DataStream& operator << ( DataStream& ds, const asset& t ){
-         return ds << t.amount << t.symbol;
+      void print()const {
+         asset::print();
+         prints("@");
+         printn(contract);
       }
-      template<typename DataStream>
-      friend DataStream& operator >> ( DataStream& ds, asset& t ){
-         return ds >> t.amount >> t.symbol;
+      extended_asset operator-()const {
+         extended_asset result(*this);
+         result.amount = -result.amount;
+         return result;
       }
+
+      friend extended_asset operator - ( const extended_asset& a, const extended_asset& b ) {
+         eosio_assert( a.symbol == b.symbol, "type mismatch" );
+         eosio_assert( a.contract == b.contract, "type mismatch" );
+         extended_asset result( asset(a.amount - b.amount, a.symbol), a.contract );
+
+         if( b.amount > 0 )
+            eosio_assert( a.amount > result.amount, "underflow" );
+         if( b.amount < 0 )
+            eosio_assert( a.amount < result.amount, "overflow" );
+
+         return result;
+      }
+
+      friend extended_asset operator + ( const extended_asset& a, const extended_asset& b ) {
+         eosio_assert( a.symbol == b.symbol, "type mismatch" );
+         eosio_assert( a.contract == b.contract, "type mismatch" );
+         extended_asset result( asset(a.amount + b.amount, a.symbol), a.contract );
+
+         return result;
+      }
+
+      EOSLIB_SERIALIZE( extended_asset, (amount)(symbol)(contract) )
+   };
+
+   struct token_identity {
+      symbol_name  symbol;
+      account_name contract;
+
+      EOSLIB_SERIALIZE( token_identity, (symbol)(contract) )
+   };
+
+   struct extended_price {
```

Comment:
is this in use?  seems like a better candidate for a more specific type header or implementation detail of a contract unless it needs standardization